dask/dataframe/groupby.py

```diff
@@ -1732,6 +1740,45 @@ def aggregate(self, arg, split_every=None, split_out=1, shuffle=None):
                 f"if pandas < 1.1.0. Pandas version is {pd.__version__}"
             )
 
+        if cardinality and (split_out is None and split_every is None):
+            if cardinality == "infer":
+                # Infer from first non-empty partition
+                _cardinality = self.obj.map_partitions(
+                    _groupby_cardinality_factor,
+                    self.by,
+                    arg,
+                    meta=self.obj.groupby(self.by).aggregate(arg),
+                    enforce_metadata=False,
+                )
+                max_trials = 5
+                for p in range(max_trials):
+                    _val = _cardinality.partitions[p].compute()
+                    if len(_val):
+                        cardinality = _val.iloc[0]
+                        break
+                if cardinality == "infer":
+                    raise RuntimeError(
+                        f"Cardinality inference failed! First {max_trials} "
+                        f"partitions are empty."
+                    )
+            elif not isinstance(cardinality, (int, float)):
+                # Unsupported option
+                raise ValueError
+
+            # Use cardinality -> split_out/every heuristics
+            _split_out = min(
+                max(
+                    int(cardinality * self.obj.npartitions * 0.66),
+                    1,
+                ),
+                self.obj.npartitions,
+            )
+            split_every = split_every or min(max(int(1.0 / cardinality), 2), 32)
```

Comment:
I think it's important to allow `split_every` to be one (cf https://github.com/dask/dask/issues/9406#issuecomment-1234875036), so that there isn't any repartitioning to fewer partitions before shuffling. I found this makes a significant difference for the case when there are fewer, larger partitions.

Otherwise, this looks like a reasonable heuristic to me.